californium-core/src/test/java/org/eclipse/californium/core/test/ErrorInjector.java

```diff
@@ -44,11 +49,13 @@ public void setErrorOnReadyToSend() {
 
 	@Override
 	public void sendRequest(final Request request) {
+		TestTools.removeMessageObservers(request, ErrorInjectorMessageObserver.class);
```

Comment:
I don't why we need to do this ? :thinking: 
(To avoid to add same injector twice ?)

Reply:
With the fix, the CON responses are sent multiple times. And that showed irritating logs.

1. transmission => 1 log line
2. transmission => 2 log lines more (from the first and the second observer).

without a implementation error, the message will have not been retransmitted, therefore that case doesn't occur in the unit test so far.

Reply:
OK. It could also be implemented as addIfNotAlreadyContained way ?

Maybe this is even something which could be add in the API as when you add observer you rarely want to add the same twice ? :thinking: 

(It's Ok if we keep this like this too)

Reply:
In this use-case, a MessageObserver replaces not the same instance but a instance of the same implementation class. I don't see, that this is a "general criteria". Also , the most application message observers are added before the message is handed over to the CoapEndoint. Therefore the "adds" are not repeated on retransmission. My feeling is, the error injector and retransmission within the unit tests are very, very special cases. It costed me some more than a hour to find out, why there are "so many duplicated" log lines while the root cause of the error wasn't fixed and the messages have been retransmitted even if the sent has failed. For that "analysis time", I thought, that fixing it for analysing future errors, is a good idea.

Reply:
I have no objection at all about removing irritating logs !  I think its a good point.

My point was more that a `addIfNotAlreadyContained` for message observer could make sense. I already experiment case where callback are called several time but this is not a so important feature and we release an RC so, again it's ok to let it like this. (that was just an idea :-))

> In this use-case, a MessageObserver replaces not the same instance but a instance of the same implementation class.

Of course, instead of create a new instance we should change the code to use only one instance of `ErrorInjectorMessageObserver` by `ErrorInjector` and so we will be in the case where we replace the same instance which seems a "more general" use case.

Reply:
The instance of `LoggingMessageObserver` is currently accessing the request/response. To reuse the instance, that must be adapted into a field. For me, remove in the `TestTools` was the solution with the fewest changes.